### ISSUE TITLE
Separation of depositor and ticket signer

### ISSUE THREAD

**Issue by yondonfu** [livepeer/protocol#252]
We should allow the account that deposits ETH to the PM contract to be different than the account that signs tickets. A depositor could specify account(s) that are allowed to sign on its behalf. See https://github.com/livepeer/LIPs/issues/13 for some benefits of this approach.

A rough sketch of what the implementation could look like:

```
struct Sender {
    ...
    mapping (address => bool) signers;
}

struct Ticket {
    ...
    address sender;
}
```

```
function isValidSenderSig(address _sender, bytes _sig, bytes32 _ticketHash) internal view returns (bool) {
    // Recover signer address from _sig and _ticketHash
    return senders[_sender].signers[signer];
}
```

**Comment by yondonfu** [comment]
One consideration here is that we need to make sure that a sender cannot invalidate a previously valid signature over a ticket by modifying its signer set after a recipient locally validates the ticket and before the recipient is able to redeem the ticket on-chain (a malicious sender could front run winning ticket redemption transactions).

We can defend against this by relying on the unlock period described in #245 - a sender can only modify its signer set after it waits through the unlock period. Recipients will be able to redeem winning tickets during this period knowing that the ticket signatures cannot be invalidated.

**Comment by dob** [comment]
One other idea that isn't fully thought through is "allowances" per signer key - much like the `approve()` method on ERC-20. The central key that manages funds could manage allowances independently of being online on many ingest machines to sign PM tickets.

**Comment by eladmallel** [comment]
Remaining TODO:

- [ ] Revocation of signers with some revocationPeriod waiting time

**Comment by eladmallel** [comment]
Brainstorming some UX aspects of revoking signers:

What's a better interface for revoking between the following options:

1. Sender calls a function called `resetSigners(address[])` and always has to supply the complete set of new signers, then needs to call `completeResetSigners` to apply the changes
2. Sender calls `requestSignersRevocation(address[])` supplying the signers to revoke, the contract marks those signers with a future block after which they are revoked

Assuming approach 2, what's an overall better solution, considering UX and security:

1. Should signers be automatically considered revoked if `currentBlock > signer.revocationBlock`
2. Or should we require sender to call another function like `completeSignersRevocation(address[])`? and if so, what do we do if sender this time sends a subset of the signers they sent prior to `requestSignersRevocation`?

@yondonfu would love to get your thoughts!

**Comment by yondonfu** [comment]
I prefer option 2 for both parts since it only requires a single transaction and the actual revocation is implicitly finalized once `currentBlock >= signer.revocationBlock`.

Something that came to mind is whether or not a sender should be able to reverse its initial decision to revoke a signer address prior to the signer address' revocation block. I'm inclined to not allow this as a sender can simply just generate a new signer address. Whether the sender approves a new signer address or reverses a previous revocation initiation an on-chain transaction will be required.

**Comment by eladmallel** [comment]
Sounds good! So you're down with only one txn to request the revocation and
have it apply automatically.

To your last point - I guess we don't really need a reset function if we
allow sender to simply approve a sender they have already revoked or
attempted to revoke.

What do you think?

**Comment by yondonfu** [comment]
>  So you're down with only one txn to request the revocation and
have it apply automatically.

Yep.

> I guess we don't really need a reset function if we
allow sender to simply approve a sender they have already revoked or
attempted to revoke.

Good point. Yeah I think that would work, `approveSigners()` would just need to make sure that an approved signer address' revocation block is set to 0 if it is not already set to 0.

**Comment by yondonfu** [comment]
Closed by #265
